File: Study Plan/Top Interview 150/0219 Contains Duplicate II/ContainsDuplicateII.cs
/*  
    Problem
    --------------------------------------------------------------------------------
    219. Contains Duplicate II
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    - Runtime 270 ms (Beats 81.83% of users with C#)
    - Memory 66.01 MB (Beats 73.53% of users with C#)
*/

public class Solution 
{
    public bool ContainsNearbyDuplicate(int[] nums, int k)
    {
        var numberAndIndex = new Dictionary<int, int>(nums.Length);

        for (var index = 0; index < nums.Length; index++)
        {
            var number = nums[index];

            if (!numberAndIndex.ContainsKey(number))
            {
                numberAndIndex.Add(number, index);
            }
            else
            {
                if (Math.Abs(numberAndIndex[number] - index) <= k)
                {
                    return true;
                }
                else
                {
                    numberAndIndex[number] = index;
                }
            }
        }

        return false;
    }
}

File: Study Plan/Top Interview 150/0219 Contains Duplicate II/ContainsDuplicateII.java
/*  
    Problem
    --------------------------------------------------------------------------------
    219. Contains Duplicate II
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    - Runtime 17 ms (Beats 89.27% of users with Java)
    - Memory 55.76 MB Beats 74.68% of users with Java)
*/

class Solution {
    public boolean containsNearbyDuplicate(int[] nums, int k) {
        HashMap<Integer,Integer> numberAndIndex = new HashMap<>(nums.length);
        for (int index = 0; index < nums.length; index++) {
            int number = nums[index];
            if (!numberAndIndex.containsKey(number)) {
                numberAndIndex.put(number, index);
            } else {
                if (Math.abs(numberAndIndex.get(number) - index) <= k) {
                    return true;
                } else {
                    numberAndIndex.put(number, index);
                }
            }
        }
        return false;
    }
}
